### PR TITLE
slate-hyperscript tests and decorations

### PR DESCRIPTION
**Pull request by jasonphillips**
This PR first adds tests to `slate-hyperscript`, covering all its basic functionality as well as its handling of selection (by cursor, and by anchor / focus).

It then adds the ability to similarly notate decorations in hyperscript, using either a syntax similar to marks (if it's a simple range within a continuous section of text), or using a `[decoration]Anchor` / `[decoration]Focus` syntax.

For the simple syntax, [see this test](https://github.com/jasonphillips/slate/blob/hyperscript-tests-and-decorations/packages/slate-hyperscript/test/decorations/single.js)

For the Anchor / Focus syntax (must supply matching key since can be multiple), [see this test](https://github.com/jasonphillips/slate/blob/hyperscript-tests-and-decorations/packages/slate-hyperscript/test/decorations/across-blocks.js)

I dove into these improvements mainly as a preparation for being able to more rigorously test decorations (I'm working on solving https://github.com/ianstormtaylor/slate/issues/1291 ), which is greatly helped when `slate-hyperscript` can easily include decorations in its readable notation.

### CODE REVIEW

**Comment by jasonphillips** [comment]
I just noticed @Soreine 's PR on slate-hyperscript https://github.com/ianstormtaylor/slate/pull/1379 -- after looking through it, however, none of those changes will conflict with these. The couple of inline tests he added can also very easily be incorporated into the nested directory of tests here upon merging these two PRs.

File: packages/slate-hyperscript/test/decorations/across-blocks.js [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  decorators: {
+    highlight: 'highlight',
+  },
+})
+
+export const input = (
+  <value>
+    <document>
+      <paragraph>
+        This is one <highlightAnchor key="a" />block.
```
Comment (Soreine) [inline packages/slate-hyperscript/test/decorations/across-blocks.js]: What are the decoration `key` used for? 
Cannot you use the decoration `key` as a way to pair the start/end of the decoration?

If not, I would suggest the following syntax:

```jsx
<highlight anchor key="a"/>
<highlight focus key="a" />
```

**Comment by jasonphillips** [comment]
```js
<highlight anchor key="a"/>
<highlight focus key="a"/>
```

Thanks, that's a much better syntax suggestion, and I've updated to match it. 

These two updated tests help illustrate why the `key` attribute is needed to distinguish overlapping versus nested decorations in complex cases:
  - [overlapping](https://github.com/jasonphillips/slate/blob/2ddebd36d5cfb32c781ce1c9b3aae1628ac4536c/packages/slate-hyperscript/test/decorations/overlapping.js) versus
  - [nested](https://github.com/jasonphillips/slate/blob/2ddebd36d5cfb32c781ce1c9b3aae1628ac4536c/packages/slate-hyperscript/test/decorations/nested.js)

**Comment by jasonphillips** [comment]
Huh, the CI build failed because of this error in upath (on the latest version of node) which suddenly hit us, unrelated to the PR:

https://github.com/anodynos/upath/issues/14

**Comment by Soreine** [comment]
Well it looks like we can still tell which decoration edges pair together,
just using the key. Following the document order, the first with key a is
the anchor, the second with key a is the focus

>

**Comment by Soreine** [comment]
@jasonphillips do you know what should be done about `upath` ?
Should we tell Travis to not use Node 10 ? Or should we tell Travis to do `yarn install --ignore-engines` as suggested in the upath issue ?

**Comment by jasonphillips** [comment]
I suppose since this is really a browser-focused library (except perhaps for the de/serialization packages that I do use on the server a bit), "officially" testing it on cutting-edge node versions in CI isn't necessarily something that matters? In which case, just setting Travis to use the prior stable Node certainly seems reasonable to me for now.

**Comment by ianstormtaylor** [comment]
@jasonphillips this is absolutely amazing. These tests look great, and the ability to model decorators with hyperscript is super smart!

I didn't realize that the two PRs would conflict, and I've already merged the other. Would you be able to resolve the conflicts and then I can merge this?

Thanks so much for your help.

**Comment by jasonphillips** [comment]
No problem -- latest master merged & conflicts resolved. I moved the 3 inline tests from the other PR into the test directories of this one.

**Comment by ianstormtaylor** [comment]
@jasonphillips thank you! Moving those is above and beyond, thank you very much.

**Pull request merged**